Игровая механика

KCherkasov opened the pull request:
Фронтенд : http://lands-dangeous.herokuapp.com/

игровая механика:
  - лобби + матчмейкинг
  - процедурно генерируемые подземелья (PvE и PvP режимы)
  - монстры + ИИ
  - фасадный класс игрового мира

ресурсная система:
  - классы персонажей, перки, ветки перков
  - роли монстров
  - процедурная генерация монстров
  - процедурная генерация предметов
  - тесты на ресурсную систему (поднятие фабрик / ассет холдеров + отдача ресурсов/PCG)

вебсокеты:
  - смарт-контроллер (обертка над сокетом + кеш для привязанного юзера)
  - Object pool для смарт-контроллеров
  - сообщения

состояния:
   - распределение хендлеров по различным состояниям игры
   - стейт-машина переработана в сервис

KCherkasov committed:
Merge pull request #2 from java-park-mail-ru/master

 DataBase support, Tests

KCherkasov committed:
-Задание к РК2 (переписать UserService.getAllUsers() на RowMapper'е;
-Подавил ряд SpellCheckInspection'ов и несколько других инспекций

KCherkasov committed:
-Задание к РК2 (переписать UserService.getAllUsers() на RowMapper'е;
-Подавил ряд SpellCheckInspection'ов и несколько других инспекций

KCherkasov committed:
-Задание к РК2 (переписать UserService.getAllUsers() на RowMapper'е;
-Подавил ряд SpellCheckInspection'ов и несколько других инспекций

KCherkasov committed:
-testGetAllUsers() fix

KCherkasov committed:
-testGetAllUsers() fix

KCherkasov committed:
gamemechanics sandbox

KCherkasov committed:
gamemechanics sandbox

KCherkasov committed:
game mechanics sandbox;
  -some cleanup and fixes;
  -adjacency lists making added;
  -moving to component-based approach

KCherkasov committed:
game mechanics sandbox;
  -some cleanup and fixes;
  -adjacency lists making added;
  -moving to component-based approach

KCherkasov committed:
game mechanics sandbox;
  -some cleanup and fixes;
  -added Property components family;
  -refactored Ability in the component-based way;
  -started working on user characters / NPCs;
  -started working on pathfinding algorithm

KCherkasov committed:
game mechanics sandbox;
  -some cleanup and fixes;
  -added Property components family;
  -refactored Ability in the component-based way;
  -started working on user characters / NPCs;
  -started working on pathfinding algorithm

KCherkasov committed:
Merge pull request #3 from java-park-mail-ru/master

Доп задание к РК2

KCherkasov committed:
game mechanics sandbox;
    -A* pathfinding class written;

KCherkasov committed:
game mechanics sandbox;
    -A* pathfinding class written;

KCherkasov committed:
game mechanics sandbox;
    -A* pathfinding class written;

KCherkasov committed:
game mechanics sandbox;
    -A* pathfinding class written;

KCherkasov committed:
game mechanics sandbox:
    -some battle-related fixes and cleanup

KCherkasov committed:
game mechanics sandbox:
    -some battle-related fixes and cleanup

KCherkasov committed:
game mechanics sandbox:
    -Battlefield refactoring

KCherkasov committed:
game mechanics sandbox:
    -Battlefield refactoring

KCherkasov committed:
game mechanics sandbox:
    - Battlefield refactoring;
    - Added interfaces for map routes, map nodes and pathfinding algos;
    - Wrote abstract character and PC classes;
    - Wrote a tileset maker for AoE abilities;

KCherkasov committed:
game mechanics sandbox:
    - Battlefield refactoring;
    - Added interfaces for map routes, map nodes and pathfinding algos;
    - Wrote abstract character and PC classes;
    - Wrote a tileset maker for AoE abilities;

KCherkasov committed:
game mechanics sandbox:
    - Wrote a MapNodeTileset class

KCherkasov committed:
game mechanics sandbox:
    - Wrote a MapNodeTileset class

KCherkasov committed:
game mechanics sandbox:
    - Wrote a field of vision estimation class

KCherkasov committed:
game mechanics sandbox:
    - Wrote a field of vision estimation class

KCherkasov committed:
game mechanics sandbox:
    - Added getOwnerID() method to the AliveEntity interface

KCherkasov committed:
game mechanics sandbox:
    - Added getOwnerID() method to the AliveEntity interface

KCherkasov committed:
game mechanics sandbox:
    - refactored CharacterClass to the interface implementation
    - refactored AliveEntity and its implementations to work with newly added interface
    - wrote a CharacterRole interface
    - wrote an Action interface
    - refactored BattleAction to be an Action implementation

KCherkasov committed:
game mechanics sandbox:
    - refactored CharacterClass to the interface implementation
    - refactored AliveEntity and its implementations to work with newly added interface
    - wrote a CharacterRole interface
    - wrote an Action interface
    - refactored BattleAction to be an Action implementation

KCherkasov committed:
game mechanics sandbox:
    - refactored AliveEntity implementations;
    - added NonPlayerCharacterRole class;
    - minor refactoring on battlefield package

KCherkasov committed:
game mechanics sandbox:
    - refactored AliveEntity implementations;
    - added NonPlayerCharacterRole class;
    - minor refactoring on battlefield package

KCherkasov committed:
game mechanics sandbox:
    - wrote javadoc for gamemechanics.interfaces and part of gamemechanics.battlefield.map.tileset packages;
    - started to write an AI class;

KCherkasov committed:
game mechanics sandbox:
    - wrote javadoc for gamemechanics.interfaces and part of gamemechanics.battlefield.map.tileset packages;
    - started to write an AI class;

KCherkasov committed:
game mechanics sandbox:
    - minor fix on NonPlayerCharacterClass;
    - wrote an AI class;

KCherkasov committed:
game mechanics sandbox:
    - minor fix on NonPlayerCharacterClass;
    - wrote an AI class;

KCherkasov committed:
game mechanics sandbox:
    - wrote first AI for monsters;
    - some refactoring on Action implementations;
    - minor refactoring on abilities;

KCherkasov committed:
game mechanics sandbox:
    - wrote first AI for monsters;
    - some refactoring on Action implementations;
    - minor refactoring on abilities;

KCherkasov committed:
Merge remote-tracking branch 'origin/develop-gamemechanics' into develop-gamemechanics

KCherkasov committed:
game mechanics sandbox:
    - added basement for backend event processing;
    - minor refactoring in Battlefield class;

KCherkasov committed:
game mechanics sandbox:
    - some more code for events and their logging;
    - barebones for PvE and PvP instances implementations;
    - minor refactoring on Action interface;

KCherkasov committed:
game mechanics sandbox:
    - AI fixup (added aggro evaluation to the behavior)

KCherkasov committed:
Merge pull request #18 from KCherkasov/gamemechanics

Костяк гейм-механики

KCherkasov committed:
game mechanics sandbox:
    - fixed bug in AI;
    - added some functionality to charcter abilities' behavior

KCherkasov committed:
game mechanics sandbox:
    - started to write abilities' behaviors
    - added a new Property implementation (based on HashSet)

Commit:
Merge pull request #2 from java-park-mail-ru/gamemechanics

Gamemechanics up

KCherkasov committed:
game mechanics sandbox (abilities behavior-related stuff):
    - refactored AggregatedAbilityAction a bit;
    - refactored Tileset and AreaEffectTileset to enable event logging;
    - wrote basic ability behavior;
    - added an EffectModel list to IngameAbility and its model;
    - refactored Ability interface to get EffectModels list from it;
    - added critical hit damage/heal multiplication constant;
    - wrote an overload for UserCharacter's getProperty(Integer, Integer) method: now it takes into accordance character's equipment, perks and effects;

KCherkasov committed:
game mechanics sandbox:
    - minor reformatting;
    - wrote events for effect application and XP/gold rewards per kill logging;
    - tuned Battlefield's event processing to enable effects application and kill rewards giving autologging;
    - wrote a helper to calculate amount of cash to give as kill reward;
    - wrote a character party container for matchmaking;
    - closed some TODOs;

KCherkasov committed:
Merge pull request #19 from KCherkasov/gamemechanics

Gamemechanics sandbox

KCherkasov committed:
Merge pull request #4 from java-park-mail-ru/gamemechanics

Gamemechanics

KCherkasov committed:
game mechanics sandbox:
    - wrote spawn point emplacement algorithm;
    - refactored AbstractInstanceModel;
    - added some functionality to the CharactersParty class;
    - refactored Instance interface and its implementations to work with CharacterParty instead of Squad;
    - started to work on PvE-instance class;
    - initialized game mechanics' resources package.

KCherkasov committed:
game mechanics sandbox:
    -wrote resource system for assets (PCG and resources for it are WIP);
    -refactored in-game asset classes to work with resources;
    -further work on instances

KCherkasov committed:
Merge remote-tracking branch 'origin/gamemechanics' into gamemechanics

KCherkasov committed:
Merge pull request #20 from KCherkasov/gamemechanics

Gamemechanics

Commit:
messages

KCherkasov committed:
game mechanics sandbox:
    - wrote AssetProvider class aggregating all non-PCG asset holders

KCherkasov committed:
game mechanics sandbox:
    - wrote PCG items factory and resources for it;
    - added Jackson annotations for serializable objects;

KCherkasov committed:
Merge pull request #21 from KCherkasov/gamemechanics

Gamemechanics (non-PCG assets and items PCG)

Commit:
connection pool, smart controller and messages up

Commit:
Merge pull request #22 from zonder129/websockets

Websockets

Commit:
Merge pull request #3 from java-park-mail-ru/gamemechanics

Gamemechanics

Commit:
no packet, only message

Commit:
Merge pull request #23 from zonder129/websockets

Websockets

KCherkasov committed:
Merge pull request #5 from java-park-mail-ru/gamemechanics

Gamemechanics

KCherkasov committed:
game mechanics sandbox:
    - big checkstyle fix;

KCherkasov committed:
Merge pull request #24 from KCherkasov/gamemechanics

Gamemechanics (checkstyle fix)

Commit:
WebSocket update

Commit:
Merge branch 'gamemechanics' into websockets

Commit:
Merge branch 'gamemechanics' into websockets

Commit:
Merge pull request #25 from zonder129/websockets

WebSocket update

KCherkasov committed:
Merge pull request #6 from java-park-mail-ru/gamemechanics

Gamemechanics

Commit:
all in new package project

Commit:
all in new package project

Commit:
priorityqueue to arraydeque in connectionPool

Commit:
little fixes

Commit:
methods update

Commit:
handleMessage from websocket to smart is ok

KCherkasov committed:
game mechanics sandbox:
    - started NPC PCG package;
    - some minor fixes in AliveEntity's implementations;
    - added random NPC role and character race getters to AssetProvider;
    - minor code reformatting

KCherkasov committed:
game mechanics sandbox:
    - started NPC PCG package;
    - some minor fixes in AliveEntity's implementations;
    - added random NPC role and character race getters to AssetProvider;
    - minor code reformatting

KCherkasov committed:
Merge remote-tracking branch 'origin/gamemechanics' into gamemechanics

KCherkasov committed:
Merge pull request #26 from KCherkasov/gamemechanics

Gamemechanics

KCherkasov committed:
Merge pull request #7 from java-park-mail-ru/gamemechanics

base repo sync

KCherkasov committed:
game mechanics sandbox:
    - wrote PCG factory for NPCs;
    - wrote special bag for monsters to store loot data and generate loot;
    - little fixups in items PCG factory;
    - added couple new affectors.

KCherkasov committed:
Merge pull request #27 from KCherkasov/gamemechanics

Gamemechanics

KCherkasov committed:
game mechanics sandbox:
    - wrote container for pending user characters' loot;
    - fixups to integrate the container from above into character parties and so on;
    - checkstyle-related fixes;
    - added item drop chance to item blueprints and loot dropping mechanics.

Commit:
message up

KCherkasov committed:
game mechanics sandbox:
    - wrote NPCs PCG factory;
    - wrote an abstract factory for all PCG content;
    - wrote PvE instances and some code for PvP ones;
    - checkstyle fixes;

KCherkasov committed:
Merge pull request #28 from KCherkasov/gamemechanics

Gamemechanics

KCherkasov committed:
Merge pull request #8 from java-park-mail-ru/gamemechanics

Gamemechanics

KCherkasov committed:
game mechanics sandbox:
    - wrote PvP instances;
    - minor refactoring of resources-related stuff;

KCherkasov committed:
game mechanics sandbox:
    - ingame lobby WIP;
    - added some properties to identify character's party and current instance;
    - added character role property and active role property to indicate character's role in squad;

KCherkasov committed:
Merge pull request #29 from KCherkasov/gamemechanics

Gamemechanics

Commit:
handlers and messages

Commit:
first commit

Commit:
Merge pull request #6 from java-park-mail-ru/gamemechanics

Gamemechanics

Commit:
not working

Commit:
error fix

Commit:
no checkstyle

Commit:
no checkstyle

Commit:
Merge pull request #31 from zonder129/websockets

With Websockets

KCherkasov committed:
Merge pull request #9 from java-park-mail-ru/gamemechanics

Gamemechanics

Commit:
remove garbage

Commit:
remove garbage

KCherkasov committed:
game mechanics sandbox:
    - ingame lobby ready;
    - added some properties to identify character's party and current instance;
    - minor CharacterParty refactoring;
    - minor ConnectionPool refactoring;

KCherkasov committed:
Merge pull request #32 from KCherkasov/gamemechanics

Gamemechanics

Commit:
Merge branch 'websockets' into gamemechanics

Commit:
Merge branch 'websockets' into gamemechanics

Commit:
Merge pull request #7 from java-park-mail-ru/gamemechanics

another merge

KCherkasov committed:
game mechanics sandbox:
    - added battle map encoding for serialization;

Commit:
some messages ready

Commit:
some messages ready

KCherkasov committed:
game mechanics sandbox:
    - wrote World class;
    - added game statistics property to user characters;

KCherkasov committed:
game mechanics sandbox:
    - wrote World class;
    - added game statistics property to user characters;

KCherkasov committed:
Merge pull request #33 from KCherkasov/gamemechanics

Gamemechanics

KCherkasov committed:
Merge remote-tracking branch 'origin/gamemechanics' into gamemechanics

# Conflicts:
#	src/main/java/project/gamemechanics/battlefield/Battlefield.java

KCherkasov committed:
Merge pull request #34 from KCherkasov/gamemechanics

Gamemechanics

KCherkasov committed:
game mechanics sandbox:
    - battle mechanics fix;

KCherkasov committed:
Merge pull request #36 from KCherkasov/gamemechanics

game mechanics sandbox:

KCherkasov committed:
game mechanics sandbox:
    - annotations for Route's implementations serialization;

KCherkasov committed:
Merge pull request #37 from KCherkasov/gamemechanics

annotations for Route's implementation

Commit:
messages update

Commit:
messages update

Commit:
Merge pull request #8 from java-park-mail-ru/gamemechanics

Gamemechanics

Commit:
Merge pull request #38 from zonder129/websockets

Messages implementations

Commit:
try websocket

Commit:
try websocket

Commit:
origins fix

Commit:
origins fix

Commit:
GameExecutor add, reset add everywhere

Commit:
GameExecutor add, reset add everywhere

Commit:
Merge pull request #39 from zonder129/websockets

Websockets

Commit:
handlers, resources add and up

Commit:
handlers, resources add and up

Commit:
Merge pull request #40 from zonder129/websockets

handlers, resources add and up

KCherkasov committed:
game mechanics sandbox:
    - resource files added;
    - resource system tests added;
    - some bugfixing;

KCherkasov committed:
Merge branch 'gamemechanics' into gamemechanics

KCherkasov committed:
Merge pull request #44 from KCherkasov/gamemechanics

Gamemechanics

KCherkasov committed:
Merge pull request #11 from java-park-mail-ru/gamemechanics

Merge pull request #44 from KCherkasov/gamemechanics

Commit:
Merge pull request #9 from java-park-mail-ru/gamemechanics

Gamemechanics

Commit:
fixes to PR

Commit:
up

KCherkasov committed:
game mechanics sandbox:
    - revived checkstyle;
    - finished resource system;
    - resource system covered with tests;
    - minor Battlefield fix

KCherkasov committed:
Merge pull request #45 from KCherkasov/gamemechanics

Gamemechanics: resources and tests

KCherkasov committed:
Merge pull request #12 from java-park-mail-ru/gamemechanics

Merge pull request #45 from KCherkasov/gamemechanics

Commit:
checkstyle fixes

Commit:
Merge branch 'gamemechanics' into websockets

Commit:
Merge branch 'gamemechanics' into websockets

Commit:
travis

Commit:
travis up

Commit:
in tests add delete and put

Commit:
Merge pull request #47 from zonder129/websockets

Websockets

KCherkasov committed:
Merge pull request #13 from java-park-mail-ru/gamemechanics

Gamemechanics

KCherkasov committed:
game mechanics sandbox:
    - minor Battlefield fix

KCherkasov committed:
Merge pull request #48 from KCherkasov/gamemechanics

Gamemechanics - stable

KSolovyev commented:
![image](https://user-images.githubusercontent.com/4122313/34206019-ce713a80-e595-11e7-921a-dd6676accc98.png)
![image](https://user-images.githubusercontent.com/4122313/34206031-d9a8ff5a-e595-11e7-82e2-40bb4e36ec08.png)
![image](https://user-images.githubusercontent.com/4122313/34206082-0753ef32-e596-11e7-8b69-c13573262519.png)

😨 😨 😨

KCherkasov commented:
Ну процентов 10-15 это ресурсы + разносортные держалки констант (конфиг-плана либо индексы пропертей и т.п.)

По объему добавлений - примерно треть - ресурсные файлы (особо itemParts.json)

KSolovyev commented:
можно ссылку на фронт в описание пул реквеста?

KCherkasov commented:
добавил ссылку, правда там сейчас могут активно вноситься правки)

KSolovyev commented:
а мультиплеер-то успеете к защите? Я сейчас не смог поиграть))

KCherkasov commented:
ну бэк готов, сейчас фронтендеры работают над наладкой месседжей на своем конце

KCherkasov commented:
бэк умеет матчмейкинг 4 типов: 
- PvE где вся группа состоит из персонажей 1 юзера (по сути TBS против ИИ),
- PvE где персонажи отряда управляются разными пользователями 
- PvP где противостоящими отрядами командуют 2 игрока (т.е. как первый мод PvE, только вместо монстров такая же группа другого игрока)
- PvP где все персонажи в отрядах принадлежат разным пользователям

(наш изначальный концепт включал только 2й и 4й моды, но на плейтесте народ потребовал добавить TBS-мод, добавили, благо для механики особой разницы нет, только в лобби разные очереди на разные моды (ну на 1й мод очереди по факту нет, сразу генерим инстанс и запускаем группу))

KSolovyev reviewed (changes requested):
хочется тестов. Сам-то код хорошо написан. Но он уже достиг того уровня сложности, когда глазами ничего не видно. Нужно хотя бы парочку основных кейсов. Ну и фронт ждем.
Сингл на раздеплоенной версии через сервер работает?

KCherkasov commented:
тесты делаем) пока можно посмотреть ресурсную систему (AssetSerializationTest в тестах, пакет project.gamemechanics.resources в коде), ее покрыл тестами целиком, подъем отдача по запросу, генерация монстров и вещей в фабрике

На какие аспекты тесты в первую очередь сделать?

KSolovyev commented:
нужны функциональные тесты покрывающие главные игровые сценарии. Все игроки заходят, все готовы, начался бой. Круто, если можно сделать упрощенную партию аля воин против скелета на карте 4х4. Тогда воин делает ход, удар, здоровье отнимается, скелет склеивает ласты, выдается лут. 

Если упростить нельзя, тогда надо подхачить механику и эти кейсы отдельно тестировать.
ход игрока, смерть персонажа, выдача лута и т.д.

KCherkasov commented:
понял
упрощения реализуемы, карта создается генератором в любой конфигурации)
бой можно как ИИ против ИИ поднять, так и эмулировать игрока, выдачу лута/экспы/голда по внутренним логам (которые нужны еще и для отдачи на фронт) можно отследить.

элемент лога - ActionResult и его имплементация (пакет project.gamemechanics.battlefield.actionresults) - аналог транзакции в бд - есть информация про инициатора, цель (может отсутствовать) и используемую абилку (тоже может не быть), и дальше перечень вызванных эвентов по шагам(элементарные изменения состояния боя, которые на фронтенде сейчас учатся парсить в отображение - например аое абилка в лист эвентов положит произошедшее на всех атакованных клетках).

Battlefield дополнительно парсит ActionResult'ы и подкладывает туда другие эвенты (например, если от атаки монстр умер, подкладывается цепочка эвентов вознаграждения (юзерам выдается золото и экспа), плюс даются "звездочки на фюзеляж" - меняется статистика (сейчас умеем собирать число побед в пвп/пве раздельно, число фрагов (лично убитых врагов) и ассистов (убитых при участии персонажа товарищем по группе) по пвп/пве отдельно, ну и плюс собираем суммарно нафармленное золото в статистику)

KCherkasov committed:
game mechanics sandbox:
    - minor fixes;
    - game scenarios tests (party making, loot transfer)

KCherkasov committed:
Merge pull request #50 from KCherkasov/gamemechanics

gamemechanics tests pt.1

KCherkasov commented:
Подлил немного тестов, тест боевки напишу завтра (точнее уже сегодня) днем

KSolovyev commented:
Приходите на РК

KCherkasov committed:
game mechanics sandbox:
    - minor fixes;
    - game scenarios tests (map generation,
    units emplacement, 1v1 fight)

KCherkasov committed:
Merge pull request #51 from KCherkasov/gamemechanics

game mechanics test
